Corrected duplicate ActorID detection in map editor

chrisharris333 opened the pull request:
Fixes #19286 and Closes #17895.  Reworked the code that was in place to detect duplicate ActorIDs.

chrisharris333 committed:
Corrected duplicate ActorID detection in map editor

chrisharris333 committed:
Corrected duplicate ActorID detection in map editor

Mailaender commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:
```diff
@@ -125,12 +125,15 @@ public ActorEditLogic(Widget widget, World world, WorldRenderer worldRenderer, D
 				{
-					if (editorActorLayer[actorId] != null)
+					if (kv.Key.ToString().ToLowerInvariant() == actorId.ToLowerInvariant() &&
+						CurrentActor.ID.ToString().ToLowerInvariant() != actorId.ToString().ToLowerInvariant())
```
ID is already a string, so I am not sure what those `ToString()` are for.

chrisharris333 replied:
Corrected and pushed.

Mailaender commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:
```diff
@@ -31,7 +31,7 @@ enum ActorIDStatus { Normal = 0, Duplicate = 1, Empty = 3 }
 		readonly LabelWidget typeLabel;
 		readonly TextFieldWidget actorIDField;
-		readonly LabelWidget actorIDErrorLabel;
+		LabelWidget actorIDErrorLabel;
```
This can stay `readonly` as that only applies to initializing and not changing it's properties.

chrisharris333 replied:
Gotcha.  Corrected this and pushed.

chrisharris333 committed:
Made corrections per reviewer Mailaender

chrisharris333 commented:
I have made the requested corrections and pushed to the bleed branch.  Is there anything else I need to do on my end?

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:
```diff
@@ -125,12 +125,15 @@ public ActorEditLogic(Widget widget, World world, WorldRenderer worldRenderer, D
 				{
-					if (editorActorLayer[actorId] != null)
+					if (kv.Key.ToLowerInvariant() == actorId.ToLowerInvariant() &&
+						CurrentActor.ID.ToLowerInvariant() != actorId.ToLowerInvariant())
```
This condition will not change inside the loop, so we can pull it out again:
```
if (!CurrentActor.ID.Equals(actorId, StringComparison.OrdinalIgnoreCase))
{
	foreach (var kv in world.Map.ActorDefinitions)
	{
		if (kv.Key.Equals(actorId, StringComparison.OrdinalIgnoreCase))
		{
			[...]
		}
	}
}
```

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:
```diff
@@ -125,12 +125,15 @@ public ActorEditLogic(Widget widget, World world, WorldRenderer worldRenderer, D
+				// Check for duplicate ActorID
+				foreach (var kv in world.Map.ActorDefinitions)
 				{
-					if (editorActorLayer[actorId] != null)
```
Is there a difference between the actors in `world.Map.ActorDefinitions` and `editorActorLayer`? I'd expect them to be the same initially, but only `editorActorLayer` getting updated while working on the map.

abcdefg30 replied:
Thinking about it, the real error here could only be the missing `!` in front of the first if, or am I missing something else?

chrisharris333 replied:
I'm very new to all of this, so please bear with me.  What you have said matches my observations.  I don't think I understand your question/concern?

pchote replied:
We definitely shouldn't be using `world.Map.ActorDefinitions` here. @abcdefg30 is correct that this is only used during the initial map load and (re)saving.

abcdefg30 replied:
@chrisharris333 my apologies. To be more clear: I was wondering if just changing `CurrentActor.ID.Equals(actorId, StringComparison.OrdinalIgnoreCase)` to `!CurrentActor.ID.Equals(actorId, StringComparison.OrdinalIgnoreCase)` and keeping the `editorActorLayer[actorId] != null` check is enough to fix the issue.

chrisharris333 replied:
@abcdefg30 I'll test that out.

chrisharris333 replied:
Doesn't work just adding the !.  @pchote Is there a list of actors that does get updated as the map gets worked on?

pchote replied:
The `if (editorActorLayer[actorId] != null)` is already doing the right thing here. You just need to make sure that `actorId` is already ToLowered.

chrisharris333 replied:
Ok, I figured out why it didn't appear to be doing anything.  Stay Tuned.

chrisharris333 replied:
So it was working as expected, but it just wasn't doing anything.  The few new lines I added do make the "Duplicate ActorID" message actually show up now, with the original detection code.

chrisharris333 replied:
@abcdefg30 and @pchote please take another look when you get the chance.  Thanks.

chrisharris333 committed:
Put most of original code back in place per abcdef30 and pchote

abcdefg30 commented:
I had to do the following changes to make this work properly: https://github.com/OpenRA/OpenRA/commit/b8fc1a5759afb0c424ac664e37df54e2fa9f2f03

My test case was creating an empty map, placing two actors and then decreasing the number of the second actor by one. Without the changes the dupe detection would never work because `ActorX` is always unequal to `actorX`.

chrisharris333 commented:
I feel like I caused more work than I solved on this, my first real contribution to an open source project.  Apologies for that.  Should I keep trying?

Additionally, is there anything else I need to do to finish this one off?

Thanks.

abcdefg30 commented:
> I feel like I caused more work than I solved on this, my first real contribution to an open source project. Apologies for that. Should I keep trying?

Don't worry, everyone needs to start somewhere. ^^

> Additionally, is there anything else I need to do to finish this one off?

Adding the changes from https://github.com/OpenRA/OpenRA/commit/b8fc1a5759afb0c424ac664e37df54e2fa9f2f03 into this PR should then cover all cases, unless I missed some other case?

Finally it would be good if you can squash all the commits of the PR before we merge it.

chrisharris333 committed:
Corrected duplicate ActorID detection in map editor

Corrected duplicate ActorID detection in map editor

Made corrections per reviewer Mailaender

Put most of original code back in place per abcdef30 and pchote

chrisharris333 committed:
Merge remote-tracking branch 'origin/bleed' into bleed

chrisharris333 commented:
@abcdefg30 - Thank you for the encouragement.  I have somehow managed to screw this up too.  I squashed the existing commits with no trouble (I think).  I couldn't see any obvious way to add your commit, so I just made the changes on my end, tested, checked in and pushed those changes - there is now a new changeset - but now I cannot figure out how to squash those commits together.  I fear I may have screwed something up.  Please let me know what I need to do to fix?

Thanks again.

Smittytron commented:
Can squash with `git rebase upstream/bleed -i`. Mark the commits that aren't the correct one (`d0d3b2f`) with a `d` for drop commit. Then force push the branch.

If you do future PRs, you will likely use `f` for fixup instead of `d`, which will merge changes into the correct existing commit.

chrisharris333 commented:
Dropped the specified commit, force pushed, now I'm getting merge conflicts?  Seems odd.  I'm not sure how to fix.  Can we ditch this PR and make a new one?  :)

Smittytron commented:
Sorry, I meant drop all commits except for the one I specified.

chrisharris333 committed:
Apply changes from b8fc1a5

chrisharris333 commented:
@Smittytron - Ok, I did that, was able to finish this time.  I'm afraid I don't know enough to tell if it is right now.  Would you please take a look?

Thank you for your help.

Smittytron commented:
I might have misread @abcdefg30's comment thinking b8fc1a5 was meant to replace everything else. If we're just trying to squash everything instead then you just need to `git rebase upstream/bleed -i` and use an `f` for fixup on every commit except the top one.

abcdefg30 commented:
> Can we ditch this PR and make a new one? :)

If you have trouble with git (which can be quite tricky to use right at times) that is also ok.

chrisharris333 commented:
I have no idea how to get this back to a sensible state.  Too much work for so few lines of code.  I'm going to close this one and start a new one.
